test/e2e/apps/statefulset.go

```diff
@@ -510,6 +510,152 @@ var _ = SIGDescribe("StatefulSet", func() {
+			for i := range pods.Items {
+				gomega.Expect(pods.Items[i].Labels[appsv1.StatefulSetRevisionLabel]).To(gomega.Equal(currentRevision), fmt.Sprintf("Pod %s/%s revision %s is not equal to currentRevision %s",
+					pods.Items[i].Namespace,
+					pods.Items[i].Name,
+					pods.Items[i].Labels[appsv1.StatefulSetRevisionLabel],
+					currentRevision))
+			}
+			newImage := NewWebserverImage
+			oldImage := ss.Spec.Template.Spec.Containers[0].Image
+
+			ginkgo.By(fmt.Sprintf("Updating stateful set template: update image from %s to %s", oldImage, newImage))
+			gomega.Expect(oldImage).ToNot(gomega.Equal(newImage), "Incorrect test setup: should update to a different image")
+			ss, err = updateStatefulSetWithRetries(ctx, c, ns, ss.Name, func(update *appsv1.StatefulSet) {
+				update.Spec.Template.Spec.Containers[0].Image = newImage
+			})
+			framework.ExpectNoError(err)
+
+			ginkgo.By("Creating a new revision")
+			ss = waitForStatus(ctx, c, ss)
+			currentRevision, updateRevision = ss.Status.CurrentRevision, ss.Status.UpdateRevision
+			gomega.Expect(currentRevision).ToNot(gomega.Equal(updateRevision), "Current revision should not equal update revision during rolling update")
+
+			ginkgo.By("Not applying an update when the partition is greater than the number of replicas")
+			for i := range pods.Items {
+				gomega.Expect(pods.Items[i].Spec.Containers[0].Image).To(gomega.Equal(oldImage), fmt.Sprintf("Pod %s/%s has image %s not equal to current image %s",
+					pods.Items[i].Namespace,
+					pods.Items[i].Name,
+					pods.Items[i].Spec.Containers[0].Image,
+					oldImage))
+				gomega.Expect(pods.Items[i].Labels[appsv1.StatefulSetRevisionLabel]).To(gomega.Equal(currentRevision), fmt.Sprintf("Pod %s/%s has revision %s not equal to current revision %s",
+					pods.Items[i].Namespace,
+					pods.Items[i].Name,
+					pods.Items[i].Labels[appsv1.StatefulSetRevisionLabel],
+					currentRevision))
+			}
+
+			ginkgo.By("ADDING FINALIZER to pod0")
+			pod0name := getStatefulSetPodNameAtIndex(0, ss)
+			pod0, err := c.CoreV1().Pods(ns).Get(ctx, pod0name, metav1.GetOptions{})
+			framework.ExpectNoError(err)
+			pod0.Finalizers = []string{"example.com/myfinalizer"}
+			pod0, err = c.CoreV1().Pods(ss.Namespace).Update(ctx, pod0, metav1.UpdateOptions{})
+			framework.ExpectNoError(err)
+			pods.Items[0] = *pod0
+			defer func() {
+				err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
+					ginkgo.By("REMOVING FINALIZER to POD0 - trying")
+					pod0, err := c.CoreV1().Pods(ss.Namespace).Get(ctx, pod0.Name, metav1.GetOptions{})
+					if err != nil {
+						ginkgo.By(fmt.Sprintf("Failed retrying to set the phase on pod0: %v", err))
+						return nil
+					}
+					pod0.Finalizers = []string{}
+					_, err = c.CoreV1().Pods(ss.Namespace).Update(ctx, pod0, metav1.UpdateOptions{})
+					if err == nil {
+						ginkgo.By("REMOVING FINALIZER to POD0 - success")
+						pods = e2estatefulset.GetPodList(ctx, c, ss)
```

Comment:
I would suggest to do asserts in the main body of the test, rather than defer

Reply:
changed it as suggested